### DISCUSSION TITLE
keep_releases not work

### DISCUSSION THREAD

**Discussion by jakmax76** [deployphp/deployer#3511] [Help needed]
- Deployer version: 7.2
- Deployment OS: linux

Set parameter:
set('keep_releases', 2); 

Not affected, and version go over the default 10 number.

**Comment by jakmax76** [comment]
sorry @antonmedv, have i write something wrong? why did you close the issue?

**Comment by Schrank** [comment]
Maybe he thought it is an old bug which was fixed in 7.2? But if you have the problem with the newest version, we should fix it :-)

**Comment by antonmedv** [comment]
I think keep_releases working as expected. We have a test for this featue. 

So we can't reproduce the issue without any more additional context.

**Comment by jakmax76** [comment]
hi @antonmedv 
this is the keep releases param in my deploy.php
set('keep_releases', 2);

this is the task

task('deploy', [
    'deploy:prepare',
    'deploy:vendors',
    'deploy:publish',
    'deploy:clear_paths',
    'deploy:cleanup'
])

deploy works as expected, but the releases increase progressively without canceling the others as I expect. They also don't stop at 10 as I read that the new default value is 10 instead of 5.

**Reply by Schrank** [reply]
Can you please add the output if `deploy -vvv`

**Reply by jakmax76** [reply]
[deploy.log](https://github.com/deployphp/deployer/files/10823374/deploy.log)

**Reply by Schrank** [reply]
Can you please overwrite the task 

    task('deploy:cleanup', function () {
        $releases = get('releases_list');
        $keep = get('keep_releases');
        $sudo = get('cleanup_use_sudo') ? 'sudo' : '';
    
        run("cd {{deploy_path}} && if [ -e release ]; then rm release; fi");
    
        if ($keep > 0) {
            foreach (array_slice($releases, $keep) as $release) {
                run("$sudo rm -rf {{deploy_path}}/releases/$release");
            }
        }
    });

and add some debugging.either keep is 0 or releases is empty?

**Reply by jakmax76** [reply]
ok, now i'm not at pc. When i deploy again i'll try it. thank you. I want to try another provider too, to see if it change.

**Reply by akosglue** [reply]
I found that I don't need 10 releases, I have just set it to 3 on deployer 7.2.0 and it works, so I also couldn't reproduce this issue.

**Reply by underdpt** [reply]
I'm also having this issue with a project. I've made this to debug:

```php
task('deploy:cleanup', function () {
    $releases = get('releases_list');
    $keep = get('keep_releases');
    $sudo = get('cleanup_use_sudo') ? 'sudo' : '';

    writeln('Keeping ' . $keep . ' releases of ' . count($releases));

    run("cd {{deploy_path}} && if [ -e release ]; then rm release; fi");

    if ($keep > 0) {
        foreach (array_slice($releases, $keep) as $release) {
            writeln("Deleting $release");
            run("$sudo rm -rf {{deploy_path}}/releases/$release");
        }
    }
});
```

This is what I get:

```cmd
.\vendor\bin\dep deploy:cleanup

task deploy:cleanup
[host] Keeping 2 releases of 0
```

So it seems that the issue is while getting the number of releases available. It seems that this is a configuration value, but should't it traverse the releases directory to get the list of releases available?

Also, it seems that the output of `dep releases` is wrong. I have releases from 150 (2023/01/30) to 235 (2023/09/19) but the output shows releases from 1 (2021/11/22) to 136 (2023/01/08).

This is a project that used deployer v6, but it was migrated to v7 on 04/2021 so it should't be related.

**Reply by underdpt** [reply]
Mmmm... taking another look at this, I've been tinkering with the `.dep` folder on this project, and found two things:

1. File `releases` hasn't been touched since 2021/09. I think this is an old file from v6 and can be safely deleted.
2. File `releases_log` seems to have it's format changed at some point, it had all values with quotation marks until some point:

```
...
{"created_at":"2022-11-10T10:27:35+0000","release_name":"130","user":"no_user","target":"production"}
{"created_at":"2022-11-11T10:17:05+0000","release_name":"131","user":"no_user","target":"production"}
{"created_at":"2022-11-15T11:23:45+0000","release_name":"132","user":"no_user","target":"production"}
{"created_at":"2022-11-17T09:34:24+0000","release_name":"133","user":"no_user","target":"production"}
{"created_at":"2022-11-18T06:40:43+0000","release_name":"134","user":"no_user","target":"production"}
{"created_at":"2022-11-19T10:31:55+0000","release_name":"135","user":"no_user","target":"production"}
{"created_at":"2022-11-20T08:10:06+0000","release_name":"136","user":"no_user","target":"production"}
{"created_at":"2022-11-21T18:05:29+0000","release_name":"137","user":"no_user","target":"production"}
{"created_at":"2022-11-23T09:45:52+0000","release_name":"138","user":"no_user","target":"production"}
{"created_at":"2022-11-24T16:46:33+0000","release_name":"139","user":"no_user","target":"production"}
{"created_at":"2022-12-16T10:52:19+0000","release_name":"140","user":"no_user","target":"production"}
{"created_at":"2022-12-16T11:03:33+0000","release_name":"141","user":"no_user","target":"production"}
{"created_at":"2022-12-16T11:08:57+0000","release_name":"142","user":"no_user","target":"production"}
{"created_at":"2022-12-21T11:32:23+0000","release_name":"143","user":"no_user","target":"production"}
{"created_at":"2022-12-22T09:34:11+0000","release_name":"144","user":"no_user","target":"production"}
{"created_at":"2022-12-29T15:59:43+0000","release_name":"145","user":"no_user","target":"production"}
{"created_at":"2023-01-04T10:40:01+0000","release_name":"146","user":"no_user","target":"production"}
{ created_at : 2023-01-25T10:47:42+0000 , release_name : 147 , user : no_user , target : production }
{ created_at : 2023-01-26T17:58:40+0000 , release_name : 148 , user : no_user , target : production }
{ created_at : 2023-01-26T18:00:17+0000 , release_name : 149 , user : no_user , target : production }
{ created_at : 2023-01-30T12:00:41+0000 , release_name : 150 , user : no_user , target : production }
{ created_at : 2023-01-30T15:26:10+0000 , release_name : 151 , user : no_user , target : production }
{ created_at : 2023-02-07T12:36:52+0000 , release_name : 152 , user : no_user , target : production }
{ created_at : 2023-02-13T08:35:00+0000 , release_name : 153 , user : no_user , target : production }
{ created_at : 2023-02-13T16:45:46+0000 , release_name : 154 , user : no_user , target : production }
{ created_at : 2023-02-17T12:13:33+0000 , release_name : 155 , user : no_user , target : production }
{ created_at : 2023-02-21T17:10:06+0000 , release_name : 156 , user : no_user , target : production }
...
```

In a newer project, there are no quotation marks on this file.

I've deleted all lines with quotation marks, and also all lines up to the first release existing on the system, but it didn't work:

```cmd
.\vendor\bin\dep deploy:cleanup

task deploy:cleanup
[host] Keeping 2 releases of 0
```

**Reply by david-windsock** [reply]
Same problem here: the `deploy:cleanup` task does nothing and the _releases_log_ file doesn't have quotes. 

Overwriting the tasks, we found that `$releases = get('releases_list');` is empty and this is why releases are not deleted. We use Windows dev machines, and the _release_log_ is not valid because of wrong shell escaping (see #3569 or #3603).

We manually fixed the _releases_log_ file (make each line a valid JSON) and worked, but the next release adds a wrong line again.

**Comment by antonmedv** [comment]
Can't reproduce. Please, debug.

**Reply by david-windsock** [reply]
Just make a deploy from Windows to a Linux machine and check the _release_log_ file.

**Comment by david-windsock** [comment]
This workaround solved the problem for us. Add to your _deploy.php_ file:
```php
desc('Fix release_log file');
task('fix:release_log', function() {
    cd('{{deploy_path}}');

    if(test('[ -f .dep/releases_log ]')) {
        $releasesLog = implode("\n", array_map(function($line) {
            if(json_decode($line) === null) {
                $line = str_replace('{ ', '{ "', str_replace(' }', '" }', $line));
                $line = str_replace(' : ', '": "', str_replace(' , ', '", "', $line));
            }

            return $line;
        }, explode("\n", run('cat .dep/releases_log'))));

        run("echo '" . str_replace("'", "'\\''", $releasesLog) . "' > .dep/releases_log");
    }
});
```

Then, add `before('deploy:cleanup', 'fix:release_log');` to fix the log before the cleanup task.
